runtime/permissions.rs

```diff
@@ -1203,13 +1201,69 @@ fn permission_prompt(message: &str) -> bool {
 
   #[cfg(not(unix))]
   fn clear_stdin() {
+    use winapi::shared::minwindef::TRUE;
+    use winapi::shared::minwindef::UINT;
+    use winapi::shared::minwindef::WORD;
+    use winapi::shared::ntdef::WCHAR;
+    use winapi::um::processenv::GetStdHandle;
+    use winapi::um::winbase::STD_INPUT_HANDLE;
+    use winapi::um::wincon::FlushConsoleInputBuffer;
+    use winapi::um::wincon::WriteConsoleInputW;
+    use winapi::um::wincontypes::INPUT_RECORD;
+    use winapi::um::wincontypes::KEY_EVENT;
+    use winapi::um::winnt::HANDLE;
+    use winapi::um::winuser::MapVirtualKeyW;
+    use winapi::um::winuser::MAPVK_VK_TO_VSC;
+    use winapi::um::winuser::VK_RETURN;
+
     unsafe {
-      let stdin = winapi::um::processenv::GetStdHandle(
-        winapi::um::winbase::STD_INPUT_HANDLE,
-      );
-      let flags =
-        winapi::um::winbase::PURGE_TXCLEAR | winapi::um::winbase::PURGE_RXCLEAR;
-      winapi::um::commapi::PurgeComm(stdin, flags);
+      let stdin = GetStdHandle(STD_INPUT_HANDLE);
+      // emulate an enter key press to clear any line buffered console characters
+      emulate_enter_key_press(stdin);
+      // read the buffered line or enter key press
+      read_stdin_line();
+      // flush the input buffer in the case of there being a buffered line
+      // and our emulated enter key press is still in the input buffer
+      flush_input_buffer(stdin);
+    }
+
+    unsafe fn flush_input_buffer(stdin: HANDLE) {
+      let result = FlushConsoleInputBuffer(stdin);
+      if result != TRUE {
+        panic!(
+          "Error flushing console input buffer: {}",
+          std::io::Error::last_os_error().to_string()
+        )
+      }
+    }
+
+    unsafe fn emulate_enter_key_press(stdin: HANDLE) {
+      // https://github.com/libuv/libuv/blob/a39009a5a9252a566ca0704d02df8dabc4ce328f/src/win/tty.c#L1121-L1131
```

Comment:
Bert pointed me in the direction of this emulated enter key press.

With the current solution the output looks like this when there is a buffered line:

![image](https://user-images.githubusercontent.com/1609021/133158155-4705ca62-2806-4b06-8ef6-246810adbd63.png)

And like this when there's not (the more common case)... note the blank line:

![image](https://user-images.githubusercontent.com/1609021/133158337-2841e7a8-71a1-45d1-abbe-97d3aa1cc9c7.png)

Is it worth adding to code ensure this blank line does not occur? I think any easy fix would be to peak at the input buffer before flushing it and if there is nothing in the input buffer it will move the console cursor up a line.

Reply:
Resolved in 7b93fbfeba6e78b39c8a1c482474911cee1d0d00 after discussion with @piscisaureus. When the console input buffer is empty (meaning a key press was emulated) the cursor will be moved up to prevent a blank line.